### PR TITLE
TMPE support in editor

### PR DESCRIPTION
**Pull request by kianzarrin**
[TrafficManager.zip]( https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=major-feature-asset-editor) from appvoyer.
### Review Process:
this is a major feature branch. PR to major feature branch need only one approval and PR from major feature branch to master need two approvals (at least one developer. the second one could be chamelon)
In previous PRs all the code for this branch has been reviewed together with some minor testing. 
This PR is about full-proof testing.
### Notes:
- `appveyor` has latest [TrafficManager.zip]( https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=major-feature-asset-editor) 
- Vehicle restriction is another PR #966. if you wish to test Vehicle restriction then merge that branch. but it should not be necessary because you can test lane serialization with lane arrows.
- The timed traffic light is not supported see #967 
- this code is compatible with Smart intersection mod (because I moved it to harmony).

### Some test scenarios that came to my mind:
- use the intersection editor to create an intersection with TMPE rules.
   - place that intersection in-game and check if TMPE rules persist.
       - try both with and without smart intersection mod.
   - use `Editor->asset editor->NEW  ` to load the intersection from loaded prefab collection and check if traffic rules persist.
   - disable the intersection from content manager then use `Editor->asset editor->LOAD` tp load the intersection from file and check if traffic rules persist. 
- load a map in the map editor. create some TMPE rules and save it
   - play the map and check if TMPE rules persist.
   - load the map and check if TMPE rules persist.

### CODE REVIEW

**Comment by thebugfixnet** [comment]
Not work as intended... not all line connectors are restored at the left junction when i load the created junction in editor again... even onedirection arrow at the left junction is pointing wrong (three directions instead of two line connector directions i used)

image

The direction arrows seems to be correct at the right junction.

**Comment by kianzarrin** [comment]
@thebugfixnet I am unable to reproduce the problem you speak of.
image

@krzychu124 also did some (i think) small testing the previous PR and found nothing wrong.

Can you still reproduce this problem? can you reproduce it with minimalistic mods (TMPE + harmony and unsub from everything else)

**Comment by Monniasza** [comment]
@kianzarrin The first post says "appvoyer" instead of "appveyor", which is incorrect.

**Comment by thebugfixnet** [comment]
Okay, i've now tested to create 10 different junctions with only these two mods active and it worked every time. Good 👍 
Don't know which of the other mods i had before could cause this issue.

**Comment by krzychu124** [comment]
> Don't know which of the other mods i had before could cause this issue.
Maybe you've still subscribed the workshop version?

**Comment by thebugfixnet** [comment]
> > Maybe you've still subscribed the workshop version?

Haha, if yes how should that one had saved half of the lane connectors? :)

**Comment by thebugfixnet** [comment]
I did an additional Test with a base set of mods i use normally when working in the editor or doing local tests. 

With the mods below it works flawless. 👍 

FineRoadAnarchy
Harmony
IntersectionMarkingTool
ModTools
MoveIt
NetworkDetective
NetworkDumpTool
NetworkSkins2
NodeController
PropAnarchy
SmartIntersectionBuilder
UnitedUI

**Comment by kianzarrin** [comment]
@aubergine10 can you please test this and if it works give me approval?

**Comment by krzychu124** [comment]
I am going to test it again, but it would be better if someone else could test this too.

**Comment by Monniasza** [comment]
> I am going to test it again, but it would be better if someone else could test this too.

What you have found?

**Comment by krzychu124** [comment]
> > I am going to test it again, but it would be better if someone else could test this too.
> 
> What you have found?

Maybe you also spend some free time testing it? We've improved build process so you can download working version of this pull request(and any other you want) using a link above and try it yourself :wink: (unsubscribe TM:PE first and copy files from archive to local mods folder)

**Comment by kvakvs** [comment]
Tested by creating an intersection.
Speed limits per lane: works
Priority signs: works
Junction settings: works
Parking per lane: works
Vehicle restriction: did not test

**Comment by kianzarrin** [comment]
@kvakvs Are you considering to run more test or approve the review?

**Comment by kvakvs** [comment]
Wanted to review the code, and got distracted during the working day. Never returned to it after. Will see today.

File: TLM/TLM/LoadingExtension.cs [hunk 1]
```diff
@@ -162,93 +168,82 @@ public class LoadingExtension : LoadingExtensionBase {
...
-
-                        IsGameLoaded = true;
-                        break;
-                    }
-
-                default: {
-                        Log.Info($"OnLevelLoaded: Unsupported game mode {mode}");
-                        return;
-                    }
+            if (BuildConfig.applicationVersion != BuildConfig.VersionToString(
```
Comment (kvakvs) [inline TLM/TLM/LoadingExtension.cs:179]: I really appreciate when big code chunks are moved out to functions with good names.
This one could be named something like `OnGameVersionMismatch` or similar?

File: TLM/TLM/TrafficManagerMod.cs [hunk 2]
```diff
@@ -53,7 +53,9 @@ public class TrafficManagerMod : IUserMod {
 
         internal bool InGameHotReload { get; set; } = false;
 
-        internal static bool InGame() => SceneManager.GetActiveScene().name == "Game";
+        internal static bool InGame() =>
```
Comment (kvakvs) [inline TLM/TLM/TrafficManagerMod.cs:56]: Add `///` doc comment explaining what this `InGame` means, does it include editor mode?

File: TLM/TLM/UI/TrafficManagerTool.cs [hunk 3]
```diff
@@ -378,6 +383,8 @@ public class TrafficManagerTool
         /// </summary>
         void DefaultRenderOverlay(RenderManager.CameraInfo cameraInfo)
         {
+            if (!LoadingExtension.PlayMode)
```
Comment (kvakvs) [inline TLM/TLM/UI/TrafficManagerTool.cs:386]: wrap with `{ }`
i didn't comment this in Harmony files, Harmony is like next level of complexity which i am happy to not see 😅

File: TLM/TLM/UI/TrafficManagerTool.cs [hunk 4]
```diff
@@ -541,6 +548,8 @@ void DefaultRenderOverlay(RenderManager.CameraInfo cameraInfo)
         }
 
         void DefaultOnToolGUI(Event e) {
+            if (!LoadingExtension.PlayMode)
```
Comment (kvakvs) [inline TLM/TLM/UI/TrafficManagerTool.cs:551]: wrap with `{ }`

File: TLM/TLM/Util/PlaceIntersectionUtil.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                oldSegments[i].MapInstanceIDs(newSegmentId: newSegmentIds[i], map: map);
+            }
+        }
+
+        public delegate void Handler(Dictionary<InstanceID, InstanceID> map);
+
+        /// <summary>
+        /// invoked when networkIDs are mapped and provides the user with dictionary of oldNetworkIds->newNetworkIds
+        /// </summary>
+        public static event Handler OnNetowrksMapped;
```
Comment (kvakvs) [inline TLM/TLM/Util/PlaceIntersectionUtil.cs:29]: typo `onnet OW rksmapped`

File: TLM/TLM/Util/PlaceIntersectionUtil.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                Asset2Data.Select(item => item.Key).ToSTR());
+
+            if (Asset2Data.TryGetValue(intersectionInfo, out var assetData)) {
+                Log.Info("PlaceIntersectionUtil.ApplyTrafficRules(): assetData =" + assetData);
+            } else {
+                Log.Info("PlaceIntersectionUtil.ApplyTrafficRules(): assetData not found (the asset does not have TMPE data)");
+                return;
+            }
+
+            var pathNetowrkIDs = assetData.PathNetworkIDs;
```
Comment (kvakvs) [inline TLM/TLM/Util/PlaceIntersectionUtil.cs:63]: same typo again `pathnet OW rkids`, also having that explicit type instead of `var` would be great

**Review by kvakvs — APPROVED** [review]

File: TLM/TLM/LoadingExtension.cs (hunk 1 above)
Comment (kianzarrin) [reply TLM/TLM/LoadingExtension.cs:179]: this is not a code I have changed and if you hide whitespaces you wowuld notice that.

Comment (kianzarrin) [reply TLM/TLM/LoadingExtension.cs:179]: but yeah I agree with you this place does need some cleanup :)

**Comment by kianzarrin** [comment]
@krzychu124 I cannot delete this protected branch. Can you please delete it for me?

**Pull request merged**